swiftide-agents/src/tasks/task.rs

```diff
@@ -0,0 +1,233 @@
+//! A `Task` facilitaties running multiple agents in sequence or in parallel.
+//!
+//! Agents can delegate work to each other, and the task keeps track of the work.
+//!
+//! A task takes a list of agents, a list of actions that the agents can take.
+//!
+//! It is also possible to hook into various lifecycle stages of the task.
+//!
+//! # Example
+//!
+//! ```no_run
+//! Task::builder()
+//!     .agents(vec![
+//!               Agent::builder().name("agent1").build()?,
+//!               Agent::builder().name("agent2").build()?
+//!             ])
+//!     .starts_with("agent1")
+//!     .with(Action::for_agent("agent1").delegates_to("agent2").and_back())
+//!     .with(Action::for_agent("agent2").can_complete())
+//!     .invoke("Do a task thing")
+//!     .await?;
+//! ```
+use crate::errors::AgentError;
+use crate::Agent;
+use derive_builder::{Builder, UninitializedFieldError};
+use std::collections::HashSet;
+use std::ops::Deref;
+use std::sync::atomic::{self, AtomicUsize};
+use std::sync::{Arc, Mutex};
+use thiserror::Error;
+use tokio::task::{AbortHandle, JoinSet};
+
+use super::action::{Action, ActionError};
+use super::running_agent::RunningAgent;
+
+// TODO:
+// - [ ] What if the agent is already running
+// - [ ] deadlockdouble check
+// - [ ] Check if possible to run in parallel (fine if not in current)
+// - [ ] Double check after all changes tokio lock variants are still needed
+// - [ ] Maybe store the abort handle optionally in the running agent?
+//
+#[derive(Builder, Clone, Debug)]
+#[builder(build_fn(skip, error = TaskBuilderError))]
+pub struct Task {
+    #[builder(field(ty = "Option<Vec<RunningAgent>>"))]
+    agents: Arc<tokio::sync::RwLock<Vec<RunningAgent>>>,
+    #[builder(field(ty = "Option<Vec<Action>>"))]
+    actions: Arc<Vec<Action>>,
+    starts_with: Arc<String>,
+    state: Arc<TaskState>,
```

Comment:
Perhaps make the task generic over a state instead that can be used in hooks etc. Not sure if this is needed.